### DISCUSSION TITLE
Routing shell calls with :! to a central floating window

### DISCUSSION THREAD

**Discussion by nikfp** [folke/noice.nvim#214] [Ideas]
Hi there, Loving Noice and TokyoNights, big fan of your work. 

I just started playing around with Noice and one thing I noticed is that when I run shell commands with `:! ---` it pops up in the message box as expected. However, it then times out and closes again. Sometimes I want to read the output of the command and this makes doing so nearly impossible. 

What would be a good starting point to possibly populate the output from integrated shell commands to a Nui popup that is both sticky until dismissed (with <cr> most likely) and also left justified instead of right justified like I have now? Not using Nvim-notify btw, I like the messages in the bottom right without.

**Comment by folke** [comment]
Unfortunately there is currently no way to know if a certain message is output from a command, but you can use `:Noice` or `:Noice last`. You can also create custom commands, to show the last 3 messages for example in a popup.

**Comment by nikfp** [comment]
I thought that might be your answer. Since Noice depends on Nui though, I had the thought last night of creating a wrapper window as a context to run a command. The use cases I have for needing the sticky window are all bound to custom keymaps and commands, so using this wrapper and then feeding command output to the buffer should be feasible. Technically this wouldn't be part of Noice, but my question for you is would it make sense to include this as an option? Something like `:NoiceSticky` followed by the command and it would automatically pipe the terminal output to the popup buffer? I can see the benefit for UI consistency, but I can also see how this isn't really a Noice problem, so much as it's my problem that I could also solve with using `:terminal` or just using Tmux like a sane vimmer would.

**Comment by folke** [comment]
I pretty miuch finished a new feature where you can use a cmdline filter in your routes.

So in your case, you'll be able to do:

```lua
    routes = {
      {
        view = "some_popup_view",
        filter = { cmdline = "^:!" },
      },
    },
```

Will probably push my changes tomorrow.

**Comment by folke** [comment]
I just added a new feature that allows command redirection. Check the [docs](https://github.com/folke/noice.nvim#%EF%B8%8F-command-redirection)!

**Reply by nikfp** [reply]
This is awesome from what I can see in the docs. I should have some time to play with it the next few days. I got something rudimentary working with Nui popups, but it could look better.
